vnext/Microsoft.ReactNative/IReactDispatcher.idl

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import "IReactPropertyBag.idl";
+
+namespace Microsoft.ReactNative {
+
+  // The delegate is used to create property value on-demand.
+  [webhosthidden]
+  delegate void ReactDispatcherCallback();
+
+  [webhosthidden]
+  interface IReactDispatcher
+  {
+    // True if dispatcher uses current thread.
+    Boolean HasThreadAccess { get; };
+
+    // Post task for the asynchronous execution.
+    void Post(ReactDispatcherCallback callback);
+  }
+
+  // Helper methods for the property bag implementation.
+  [webhosthidden]
+  static runtimeclass ReactDispatcherHelper
+  {
+    // Get or create IReactDispatcher for the current UI thread.
+    static IReactDispatcher UIThreadDispatcher{ get; };
+
+    // Get name of the UIThreadDispatcher property.
+    static IReactPropertyName UIThreadDispatcherProperty();
```

Comment:
Same here. Not sure we should expose what we consider to be the UI thread yet. People can already do this manually with the property bag, right?

Reply:
If we do want to directly expose a thread to the consumers and its available in all cases, maybe we should expose it on the context + InstanceSettings without letting people know it's in the bag? From the outside observer perspective, ReactHost in Office using the property bag for queues felt like it was hiding something meant to be explicit.

Reply:
Good point. We may do it in future. For now I would prefer it to be in the property bag. There are several reasons:
- We want to associate a queue with a native module. The best way to do it is to associate property ID with the native module. This property ID can be used to retrieve the real queue before we make instance of the module. It is the main reason why we are doing it in Office for Reka services.
- I do not want the ReactContext become too big or too fat. E.g. today it depends on XAML for view events and it already creates a lot of issues. Ideally, the context must be as small as possible. Its main goal is to be a central point to access shared data and resources from native modules. Ideally no preference must be given to any specific interface or component.

---
In reply to: [423973960](https://github.com/microsoft/react-native-windows/pull/4877#discussion_r423973960) [](ancestors = 423973960)

Reply:
There must be a way to put something in the property bag.  In simple UWP 'green' applications we already do it in the ReactApplication. But for win32 XAML islands, we cannot use it and we must provide ability:
- What to put as a UI dispatcher in the property bag - UIThreadDispatcher property.
- How to put it in the property bag - use the UIThreadDispatcherProperty to store the UIThreadDispatcher.

---
In reply to: [423971383](https://github.com/microsoft/react-native-windows/pull/4877#discussion_r423971383) [](ancestors = 423971383)